autopilot/autopilot.go

```diff
@@ -8,24 +8,53 @@ import (
 	"go.sia.tech/renterd/bus"
 	"go.sia.tech/renterd/hostdb"
 	"go.sia.tech/renterd/internal/consensus"
+	rhpv2 "go.sia.tech/renterd/rhp/v2"
+	"go.sia.tech/renterd/wallet"
 	"go.sia.tech/renterd/worker"
+	"go.sia.tech/siad/types"
 )
 
 type Store interface {
 	Tip() consensus.ChainIndex
+	Synced() bool // TODO: should be added to `Tip` | should be moved over to Bus
+
 	Config() Config
 	SetConfig(c Config) error
 }
 
+// TODO: should be defined in bus package, should be the interface of the client
 type Bus interface {
+	// wallet
+	WalletBalance() (types.Currency, error)
+	WalletAddress() (types.UnlockHash, error)
+	WalletTransactions(since time.Time, max int) ([]wallet.Transaction, error)
+	WalletFund(txn *types.Transaction, amount types.Currency) ([]types.OutputID, []types.Transaction, error)
+	WalletDiscard(txn types.Transaction) error
+	WalletSign(txn *types.Transaction, toSign []types.OutputID, cf types.CoveredFields) error
+
+	// hostdb
 	AllHosts() ([]hostdb.Host, error)
 	Hosts(notSince time.Time, max int) ([]hostdb.Host, error)
+	Host(hostKey consensus.PublicKey) (hostdb.Host, error)
 	RecordHostInteraction(hostKey consensus.PublicKey, hi hostdb.Interaction) error
+
+	// contracts
+	AddContract(c rhpv2.Contract) error
+	RenewableContracts(renewWindow uint64) ([]bus.Contract, error)
+	AcquireContractLock(types.FileContractID) (types.FileContractRevision, error)
+	ReleaseContractLock(types.FileContractID) error
+
+	// contractsets
+	SetHostSet(name string, hosts []consensus.PublicKey) error
 	HostSetContracts(name string) ([]bus.Contract, error)
 }
 
 type Worker interface {
 	RHPScan(hostKey consensus.PublicKey, hostIP string) (worker.RHPScanResponse, error)
+	RHPPrepareForm(renterKey consensus.PrivateKey, hostKey consensus.PublicKey, renterFunds types.Currency, renterAddress types.UnlockHash, hostCollateral types.Currency, endHeight uint64, hostSettings rhpv2.HostSettings) (types.FileContract, types.Currency, error)
+	RHPPrepareRenew(contract types.FileContractRevision, renterKey consensus.PrivateKey, hostKey consensus.PublicKey, renterFunds types.Currency, renterAddress types.UnlockHash, hostCollateral types.Currency, endHeight uint64, hostSettings rhpv2.HostSettings) (types.FileContract, types.Currency, types.Currency, error)
+	RHPForm(renterKey consensus.PrivateKey, hostKey consensus.PublicKey, hostIP string, transactionSet []types.Transaction) (rhpv2.Contract, []types.Transaction, error)
+	RHPRenew(renterKey consensus.PrivateKey, hostKey consensus.PublicKey, hostIP string, contractID types.FileContractID, transactionSet []types.Transaction, finalPayment types.Currency) (rhpv2.Contract, []types.Transaction, error)
```

Comment:
Is there a specific reason for these things being 2 step operations?

Reply:
`RHPPrepareForm` and `RHPPrepareRenew` are essentially helper functions. Technically, we don't *have* to use them -- we could just call the corresponding Go functions directly -- but someone else writing an autopilot, particularly in a language other than Go, would find it a lot easier to use these.